### PR TITLE
Replace run_compiler with RunCompiler builder pattern

### LINKED ISSUE #77286: Create a RunCompiler builder to replace the run_compiler function
This argument list is slowly getting out of hand. Maybe we should create a `RunCompiler` builder pattern struct that doesn't require most calls to use `None, None, None`?

Anyway, pre-existing, if you rather open an issue about it than fix it here, I'm fine with merging this PR

_Originally posted by @oli-obk in https://github.com/rust-lang/rust/pull/76474#r495733723_

### PR DESCRIPTION
**Pull request by lakshya-sky**
Fixes #77286. Replaces rustc_driver:run_compiler with RunCompiler builder pattern.

### CODE REVIEW

**Comment by Xanewok** [comment]
It seems you're adding `target/rls` directory - this is the build artifact/cache directory created when running RLS locally. Please don't include that in the PR, thanks :slightly_smiling_face:

**Comment by jyn514** [comment]
@dash2507 for help with git, take a look at https://rustc-dev-guide.rust-lang.org/git.html

**Review by bjorn3 — APPROVED** [review]
LGTM after removing the compiler/rustc_driver/target dir. I don't have bors permission though.

**Comment by jyn514** [comment]
@dash2507 do you mind squashing the git history so rust doesn't have thousands of build files in the git history? It will make the repo even slower to download for everyone.

https://rustc-dev-guide.rust-lang.org/git.html#advanced-rebasing

File: src/tools/clippy/src/driver.rs [hunk 1]
```diff
@@ -368,22 +363,26 @@ pub fn main() {
...
             exit(0);
         }
 
         let should_describe_lints = || {
             let args: Vec<_> = env::args().collect();
-            args.windows(2)
-                .any(|args| args[1] == "help" && matches!(args[0].as_str(), "-W" | "-A" | "-D" | "-F"))
+            args.windows(2).any(|args| {
+                args[1] == "help" && matches!(args[0].as_str(), "-W" | "-A" | "-D" | "-F")
+            })
```
Comment (bjorn3) [inline src/tools/clippy/src/driver.rs:385]: Please revert unrelated formatting changes and squash all your commits.

File: src/tools/clippy/src/driver.rs [hunk 2]
```diff
@@ -369,7 +369,7 @@ pub fn main() {
         // Setting RUSTC_WRAPPER causes Cargo to pass 'rustc' as the first argument.
         // We're invoking the compiler programmatically, so we ignore this/
         let wrapper_mode = orig_args.get(1).map(Path::new).and_then(Path::file_stem) == Some("rustc".as_ref());
-
+
```
Comment (jyn514) [inline src/tools/clippy/src/driver.rs:372]: You have trailing whitespace here, tidy will complain.
```suggestion

```

File: src/tools/clippy/src/driver.rs [hunk 3]
```diff
@@ -369,7 +369,6 @@ pub fn main() {
         // Setting RUSTC_WRAPPER causes Cargo to pass 'rustc' as the first argument.
         // We're invoking the compiler programmatically, so we ignore this/
         let wrapper_mode = orig_args.get(1).map(Path::new).and_then(Path::file_stem) == Some("rustc".as_ref());
-
```
Comment (bjorn3) [inline src/tools/clippy/src/driver.rs:372]: Please keep this line. @jyn514 was talking about the extra space you added on this line.

Comment (dash2507) [reply src/tools/clippy/src/driver.rs:372]: Sorry, but I tried a few times to remove the whitespace. but couldn't get rid of it.

Comment (bjorn3) [reply src/tools/clippy/src/driver.rs:372]: What editor are you using?

Comment (dash2507) [reply src/tools/clippy/src/driver.rs]: Neovim, never mind figured it out, removed whitespace.

**Review by bjorn3 — APPROVED** [review]
You may want to edit the commit message to be more descriptive. For example you could use just the PR title. Looking at https://forge.rust-lang.org/index.html it seems that the no tool breakage week has ended yesterday.

**Comment by Mark-Simulacrum** [comment]
This was backported into 1.48 (approved by me) in #79075 as a dependency of #78775.
